### DISCUSSION TITLE
Self-hosted runner job queue API/UI

### DISCUSSION THREAD

**Discussion by joshlk** [community/community#48708] [Actions]
### Select Topic Area

Product Feedback

### Body

It would be great to have an API/UI to query what jobs are waiting for a self-hosted runner or group/label of runners across all repos

**Comment by jsoref** [comment]
I think that https://github.com/organizations/$ORGANIZATION/settings/actions/hosted-runners provides a UI...

**Reply by joshlk** [reply]
It shows active jobs but not a job queue

**Reply by jsoref** [reply]
Ah. The queue probably, technically, doesn't work the way you want. Namely each queued job doesn't really know where it wants to go. 

I believe there are essentially four things that might handle a job:
* GitHub hosted runners
* Self-hosted repository runners
* Self-hosted organization runners
* Self-hosted Enterprise runners

Something I learned recently is that runner names don't have to be particularly unique in that one can reuse a name across these different levels.

An organization, for instance, can't ask if a job will be taken by a runner in its pools, because it doesn't know. A runner could appear at the repository or Enterprise level and take the job instead.

Unfortunately, that means, you'd have to ask each repository about its own queued jobs.

**Reply by jsoref** [reply]
There are two other things...
* [GitHub-hosted larger runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners) (organization)
* [GitHub-hosted larger runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners) (enterprise)

**Comment by FWBlueWizard** [comment]
> It would be great to have an API/UI to query what jobs are waiting for a self-hosted runner or group/label of runners across all repos

Why is this not a more ready feature? Or give us an API to see what jobs are waiting to be run on anything listed as 'self-hosted'? 
Anyone see this mentioned somewhere else? Really could use this as insight as to how much load is in demand on self-hosted runners.

**Comment by sbrakl** [comment]
@jsoref We are there enterprise custom and have same requirement. We need a API to see incoming runner request for the GitHub Enterprise and way to slice and dice per GitHub organization, GitHub runners labels it's asking and so on.
Can we collate data somehow from the GitHub organization webhook?

**Reply by jsoref** [reply]
Maybe? (It probably depends one whether you're trying to collect "waiting" or "dispatched" -- for the former, probably not, at least based on my possibly poor understanding?)

If you're looking for requests as they are dispatched (as opposed to as they're made and have not yet been dispatched), you might be able to just use: https://github.com/actions/runner/blob/main/docs/adrs/1751-runner-job-hooks.md (I have never seen it until now)

Note:
1. I don't work for GitHub
2. I don't have access to any GitHub Enterprises (I'm a minor member in one, but not one such that I could see any hook content)

If I were you, I'd do one of the following:
* open an enterprise ticket
* talk to my CSR (I do not hesitate to open tickets to my vendors / reach out to my CSRs)
* [_it sounds like fun, so I'd almost certainly start by_] collect[ing] some hook events for at least a few hours and then scour the data to see if anything interesting pops up 😄 [-- _but, my guess is that I wouldn't find anything._]
 * visit:
https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/monitoring-your-current-jobs#viewing-queued-jobs-in-your-organization-or-enterprise and either
    * use the 👎 and explain that the [`Viewing queued jobs in your organization or enterprise`](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/monitoring-your-current-jobs#viewing-queued-jobs-in-your-organization-or-enterprise) entry is useless in the feedback, or
    * open a ticket in https://github.com/github/docs/issues/new?assignees=&labels=content&projects=&template=improve-existing-docs.yaml referencing https://github.com/github/docs/blob/main/content/actions/using-github-hosted-runners/about-github-hosted-runners/monitoring-your-current-jobs.md and suggest that the section be removed or replaced with content that says "you really can't do this, but if you want to follow the same steps as the above you can see how full your queue is at this level".

Note that I file lots of tickets to the docs repository, often even for things like this (that I don't use/need), but I have more than enough open tickets at this time and I don't feel like being that belligerent/snarky there at this time, although the content they have does seem worthy of such a thing. -- I'm much more likely to use a ticketing/CSR for this.

**Comment by AmilaDevops** [comment]
Thanks for your efforts on-behalf of the community @jsoref  so after all your research could I know weather can I have any Github API from Github side to see what jobs are waiting to be run on anything listed as 'self-hosted' runner ?
